### DISCUSSION TITLE
Optional web API polyfills in @remix-run/node

### DISCUSSION THREAD

**Discussion by isker** [remix-run/remix#5858] [Proposals]
Hi.  I'm trying to convert [a project](https://github.com/isker/neogrok) to remix for the first time.  Everything's going pretty well so far, but one thing about remix's node runtime has stood out to me: the web API polyfills.  Things like fetch, FormData, the web streams API, etc.

Modern nodes [provide implementations](https://nodejs.org/api/globals.html) of every web API remix polyfills on the global object, but remix [unconditionally sets its polyfills on the global object](https://github.com/remix-run/remix/blob/a0341c97c19784a75a05124de6186ec09ca47541/packages/remix-node/globals.ts#L45-L62), blowing the native implementations away.  I'd prefer to use the native ones as they are less liable to do things like [crash the server when fetches are aborted](https://github.com/remix-run/remix/issues/4993) :grimacing:.  Especially the undici fetch implementation that ships in node is benchmarked thoroughly and achieves very high performance.

I'm wondering if there's any path toward remix's node polyfills becoming explicitly optional, conditional on node version, conditional on being missing from the global object, etc.  This is perhaps complicated by the fact that the node runtime [exports many of its polyfilled implementations](https://github.com/remix-run/remix/blob/a0341c97c19784a75a05124de6186ec09ca47541/packages/remix-node/index.ts#L13).  I'm not sure this is possible without creating an entirely new runtime, new adapters, etc.

Thanks!

**Comment by isker** [comment]
For something to compare to, you can see that next [only conditionally polyfills fetch](https://github.com/vercel/next.js/blob/66b346ed236da291176f002326e88ceab3da3c10/packages/next/src/server/node-polyfill-fetch.ts#L4).

**Comment by aniravi24** [comment]
I'm curious about this as well, because I'm actually running into issues that appear to be caused by the polyfills. The open PRs and issues about the custom polyfills also have been unresolved for a long time.

For example, I'm unable to use `AbortController` correctly.
See issue [here](https://github.com/remix-run/remix/issues/3971)

**Comment by MichaelDeBoey** [comment]
I created a PR to implement this: https://github.com/remix-run/remix/pull/6259

**Comment by MichaelDeBoey** [comment]
We just published version `1.17.1-pre.0` which includes #6562 (a possible fix). If you'd like to take it for a test run please try it out and let us know what you think!

**Comment by KyleAMathews** [comment]
Undici is still broken on 1.17.1 (edit: looks like in general Remix isn't ready for Node 18/20)

```
    throw new TypeError(`${context} is not a ReadableStream.`);
          ^
TypeError: First parameter has member 'readable' that is not a ReadableStream.
```

**Reply by jorisre** [reply]
I'm encountering the same error too. Has anyone discovered a solution?

**Reply by justinhenricks** [reply]
Also see the same issue when trying to use Youtube web loader from langchain

```
const loader = YoutubeLoader.createFromUrl("https://youtu.be/bZQun8Y4L2A", {
  language: "en",
  addVideoInfo: true,
});

const docs = await loader.load();
```
